nixos/modules/services/web-apps/discourse.nix

```diff
@@ -0,0 +1,471 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.discourse;
+  dataDir = "/var/lib/discourse";
+  user = "discourse";
+  group = "discourse";
+  baseUrl = "${if cfg.web.https then "https" else "http"}://${cfg.hostName}";
+
+  configGenerator = c: concatStrings (flip mapAttrsToList c (key: val: "${key} = ${val}\n"));
+  mainConfig = pkgs.writeText "discourse.conf" (configGenerator cfg.settings);
+
+  webConfig = pkgs.writeText "site_settings.yml" (builtins.toJSON cfg.web.settings);
+
+  dcEnv = {
+    RAILS_ENV = "production";
+    RAILS_ROOT = dataDir;
+    UNICORN_WORKERS = toString cfg.web.workers;
+  };
+
+  mailEnv = {
+    DISCOURSE_API_USER = cfg.mail.apiUser;
+    DISCOURSE_API_KEY = cfg.mail.apiKeyFile;
+    DISCOURSE_BASE_URL = baseUrl;
+  };
+
+  dcServiceConfig = {
+    Type = "simple";
+    Restart = "always";
+
+    User = user;
+    Group = group;
+
+    StateDirectory = "discourse";
+    WorkingDirectory = dataDir;
+    ProtectHome = true;
+    ProtectSystem = "full";
+    ProtectKernelTunables = true;
+    ProtectKernelModules = true;
+    ProtectControlGroups = true;
+  };
+
+  discourse-bundler = pkgs.stdenv.mkDerivation {
+    name = "discourse-bundler";
+    buildInputs = [ pkgs.makeWrapper ];
+    dontBuild = true;
+    dontUnpack = true;
+    installPhase = ''
+      makeWrapper ${cfg.package.rubyEnv}/bin/bundler $out/bin/discourse-bundler \
+        ${concatStrings (mapAttrsToList (name: value: "--set ${name} '${value}' ") dcEnv)} \
+        --set PATH '${lib.makeBinPath [ pkgs.coreutils pkgs.procps pkgs.gnugrep pkgs.nodePackages.uglify-js pkgs.which pkgs.brotli pkgs.gzip ]}:$PATH'
+    '';
+  };
+in
+{
+  ###### interface
+
+  options.services.discourse = with types; {
+    enable = mkEnableOption "Enable Discourse forum";
+
+    hostName = mkOption {
+      type = str;
+      description = "Hostname running this forum.";
+      example = "forum.example.org";
+    };
+
+    package = mkOption {
```

Comment:
Do you expect users to choose between packages? If not I'd drop the option, you can always use overlay.

Reply:
I think this could be handy to set a package with plugins at some point.